### PR TITLE
[SYCL-MLIR] Use `sycl.id.constructor`

### PR DESCRIPTION
**Pull request by whitneywhtsang**

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
@victor-eds Something wrong with `SYCLIDConstructorOp` lowering?
The tests below failed with `error: failed to legalize operation 'builtin.unrealized_conversion_cast' that was explicitly marked illegal`.
```
Failed Tests (2):
  cgeist :: Verification/sycl/stream-copy.cpp
  cgeist :: Verification/sycl/stream-triad.cpp
```

**Comment by victor-eds** [comment]
> @victor-eds Something wrong with `SYCLIDConstructorOp` lowering?

It was polygeist's lowering. In #10151, I allow these operations in `-convert-polygeist-to-llvm` and add a pass to erase them afterwards.

File: mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp [hunk 1]
```diff
@@ -33,17 +33,10 @@ SYCLRangeGetOp sycl::createSYCLRangeGetOp(Type resTy,
 TypedValue<MemRefType> sycl::constructSYCLID(IDType idTy,
```
Comment (whitneywhtsang) [inline mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp:33]: rename constructSYCLID to createSYCLIDConstructorOp

**Comment by etiotto** [comment]
Any impact on the SYCL-Bench performance ?

File: polygeist/test/polygeist-opt/sycl/kernel_disjoint_specialization.mlir [hunk 2]
```diff
@@ -105,48 +97,34 @@ gpu.module @device_func {
...
 
   // COM: Version with condition: [[ACC1_END]] <= [[ACC2_BEGIN]] || [[ACC1_BEGIN]] >= [[ACC2_END]].
-  // CHECK:         [[ACC2_BEGIN:%.*]] = sycl.accessor.subscript %arg1[%alloca_10] : (memref<?x!sycl_accessor_2_f32_w_gb>, memref<1x!sycl_id_2_>) -> memref<?xf32, 1>
-  // CHECK:         [[ACC2_END:%.*]] = sycl.accessor.subscript %arg1[%alloca_19] : (memref<?x!sycl_accessor_2_f32_w_gb>, memref<1x!sycl_id_2_>) -> memref<?xf32, 1>
+  // CHECK:         [[ACC2_BEGIN:%.*]] = sycl.accessor.subscript %arg1[%8] : (memref<?x!sycl_accessor_2_f32_w_gb>, memref<1x!sycl_id_2_>) -> memref<?xf32, 1>
+  // CHECK:         [[ACC2_END:%.*]] = sycl.accessor.subscript %arg1[%14] : (memref<?x!sycl_accessor_2_f32_w_gb>, memref<1x!sycl_id_2_>) -> memref<?xf32, 1>
   // CHECK-DAG:     [[ACC1_END_PTR:%.*]] = "polygeist.memref2pointer"([[ACC1_END]]) : (memref<?xf32, 1>) -> !llvm.ptr<1>
   // CHECK-DAG:     [[ACC2_BEGIN_PTR:%.*]]  = "polygeist.memref2pointer"([[ACC2_BEGIN]]) : (memref<?xf32, 1>) -> !llvm.ptr<1>
-  // CHECK-NEXT:    %22 = llvm.icmp "ule" [[ACC1_END_PTR]], [[ACC2_BEGIN_PTR]] : !llvm.ptr<1>
+  // CHECK-NEXT:    %18 = llvm.icmp "ule" [[ACC1_END_PTR]], [[ACC2_BEGIN_PTR]] : !llvm.ptr<1>
```
Comment (jopperm) [inline polygeist/test/polygeist-opt/sycl/kernel_disjoint_specialization.mlir:122]: Nit: Pattern-match values here (and in the other tests you're modifying).

**Review by jopperm — APPROVED** [review]

File: polygeist/test/polygeist-opt/sycl/detect-reduction-accessor-versioning.mlir [hunk 3]
```diff
@@ -25,30 +25,22 @@
...
+// CHECK-NEXT:    memref.store %4, %alloca_8[%c0_9] : memref<1x!sycl_range_1_>
+// CHECK-NEXT:    %c1_10 = arith.constant 1 : index
+// CHECK-NEXT:    %c0_i32 = arith.constant 0 : i32
+// CHECK-NEXT:    %5 = sycl.range.get %alloca_8[%c0_i32] : (memref<1x!sycl_range_1_>, i32) -> index
+// CHECK-NEXT:    %6 = sycl.id.constructor(%5) : (index) -> memref<1x!sycl_id_1_>
+// CHECK-NEXT:    [[ARG0_END:%.*]] = sycl.accessor.subscript %arg0[%6] : (memref<?x!sycl_accessor_1_i32_rw_gb, 4>, memref<1x!sycl_id_1_>) -> memref<?xi32, 1>
 
-// CHECK:         [[ARG1_BEGIN:%.*]] = sycl.accessor.subscript %arg1[%alloca_18]
-// CHECK:         [[ARG1_END:%.*]] = sycl.accessor.subscript %arg1[%alloca_25]
+// CHECK:         [[ARG1_BEGIN:%.*]] = sycl.accessor.subscript %arg1[%8]
```
Comment (sommerlukas) [inline polygeist/test/polygeist-opt/sycl/detect-reduction-accessor-versioning.mlir:42]: Can we include the definition of these values in the test? Maybe use matching for them?

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h [hunk 4]
```diff
@@ -37,8 +37,10 @@ SYCLRangeGetOp createSYCLRangeGetOp(Type resTy, TypedValue<MemRefType> range,
                                     Location loc);
 
 /// Construct sycl.id with id type \p idTy and indexes \p indexes.
-TypedValue<MemRefType> constructSYCLID(IDType idTy, ArrayRef<Value> indexes,
-                                       OpBuilder builder, Location loc);
+TypedValue<MemRefType> createSYCLIDConstructorOp(IDType idTy,
```
Comment (sommerlukas) [inline mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:40]: Probably a matter of personal preference, but I liked the old name (`constructSYCLID`) of the function better. I'll leave it up to you to decide on the name though.

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h [hunk 5]
```diff
@@ -37,8 +37,10 @@ SYCLRangeGetOp createSYCLRangeGetOp(Type resTy, TypedValue<MemRefType> range,
...
+                                                 ArrayRef<Value> indexes,
```
Comment (victor-eds) [inline mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:41]:
```suggestion
                                                 ValueRange indexes,
```
NIT

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h (hunk 4 above)
Comment (victor-eds) [reply mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:40]: TBH, I also like the old name better, but this is more inline with what we can find in [`llvm::IRBuilderBase`](https://llvm.org/doxygen/classllvm_1_1IRBuilderBase.html). Up to Whitney, I guess.

Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:40]: I will keep the changed function, as it is more inline with `IRBuilderBase`, and also other functions we have in this file.

File: polygeist/test/polygeist-opt/sycl/detect-reduction-accessor-versioning.mlir (hunk 3 above)
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/detect-reduction-accessor-versioning.mlir:42]: How about changing them to `{{.*}}`? They were intentionally not included to simplify the checks, as they are the same codegen the get the begin and end of an accessor.

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h (hunk 5 above)
Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:41]: done.

File: polygeist/test/polygeist-opt/sycl/kernel_disjoint_specialization.mlir (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/kernel_disjoint_specialization.mlir]: done.

**Comment by whitneywhtsang** [comment]
> Any impact on the SYCL-Bench performance ?

Verified there is no impact on the SYCL-Bench performance.

**Review by sommerlukas — APPROVED** [review]

**Pull request merged**